File: internal/cron/telegram_cron.go
package cron

// telegram_cron.go — Port dari src/server/jobs/telegram-cron.ts
//
// Two jobs:
// 1. telegram_backup — creates DB backup and sends to Telegram (dynamic schedule from settings)
// 2. telegram_health — sends comprehensive health report to Telegram (every hour)

import (
	"bytes"
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"strings"
	"time"

	"github.com/rs/zerolog/log"

	"github.com/s4lfanet/salfanet-radius-go/internal/db/models"
)

const telegramBackupDir = "/var/www/salfanet-radius/backups"

// jobTelegramBackup creates a DB backup and sends it to Telegram.
func (s *Scheduler) jobTelegramBackup() {
	h := s.startHistory("telegram_backup")
	defer func() { s.completeHistory(h, recover()) }()

	// Deduplication guard: skip if another run started within 5 minutes
	fiveMinAgo := time.Now().Add(-5 * time.Minute)
	var recent models.CronHistory
	if s.db.Where("jobType = ? AND status IN ? AND startedAt > ?",
		"telegram_backup", []string{"running", "success"}, fiveMinAgo).
		Order("startedAt desc").First(&recent).Error == nil {
		log.Info().Msg("cron: telegram_backup skipped (duplicate)")
		s.finishHistory(h, "Skipped — duplicate run")
		return
	}

	// Get Telegram settings
	var settings models.TelegramBackupSettings
	if err := s.db.Where("enabled = true").Order("createdAt desc").First(&settings).Error; err != nil {
		s.finishHistory(h, "Telegram backup disabled, skipped")
		return
	}

	// Create backup directory
	if err := os.MkdirAll(telegramBackupDir, 0755); err != nil {
		s.failHistory(h, fmt.Errorf("mkdir: %w", err))
		return
	}

	// Create DB dump
	now := time.Now()
	filename := fmt.Sprintf("backup_%s.sql.gz", now.Format("20060102_150405"))
	fullPath := filepath.Join(telegramBackupDir, filename)

	if err := telegramMysqlDump(fullPath); err != nil {
		s.failHistory(h, fmt.Errorf("mysqldump: %w", err))
		return
	}

	var filesize int64
	if info, err := os.Stat(fullPath); err == nil {
		filesize = info.Size()
	}

	// Send to Telegram
	caption := fmt.Sprintf("🗄 Auto backup — %s\nSize: %d KB", filename, filesize/1024)
	if err := telegramSendDocument(settings.BotToken, settings.ChatId, fullPath, caption); err != nil {
		s.failHistory(h, fmt.Errorf("send to telegram: %w", err))
		return
	}

	// Record backup history
	hist := models.BackupHistory{
		ID:       fmt.Sprintf("%d", now.UnixNano()),
		Filename: filename,
		Filepath: &fullPath,
		Filesize: filesize,
		Type:     "auto",
		Status:   "success",
		Method:   "telegram",
	}
	s.db.Create(&hist)

	// Cleanup old backups
	if settings.KeepLastN > 0 {
		var oldBackups []models.BackupHistory
		s.db.Where("type = ?", "auto").Order("createdAt desc").Find(&oldBackups)
		if len(oldBackups) > settings.KeepLastN {
			for _, old := range oldBackups[settings.KeepLastN:] {
				if old.Filepath != nil {
					os.Remove(*old.Filepath)
				}
				s.db.Delete(&old)
			}
		}
	}

	s.finishHistory(h, fmt.Sprintf("Backup sent to Telegram: %s", filename))
	log.Info().Str("file", filename).Msg("cron: telegram_backup done")
}

// jobTelegramHealth sends a comprehensive health report to Telegram.
func (s *Scheduler) jobTelegramHealth() {
	h := s.startHistory("telegram_health")
	defer func() { s.completeHistory(h, recover()) }()

	// Get Telegram settings
	var settings models.TelegramBackupSettings
	if err := s.db.Where("enabled = true").Order("createdAt desc").First(&settings).Error; err != nil {
		s.finishHistory(h, "Telegram health check disabled, skipped")
		return
	}

	// Gather health data
	health := s.gatherHealth()

	// Build report
	report := formatHealthReport(health)

	// Send to Telegram
	chatId := settings.ChatId
	if settings.HealthTopicId != nil && *settings.HealthTopicId != "" {
		chatId = chatId + ":" + *settings.HealthTopicId
	}
	if err := telegramSendMessage(settings.BotToken, chatId, report); err != nil {
		s.failHistory(h, fmt.Errorf("send health to telegram: %w", err))
		return
	}

	s.finishHistory(h, fmt.Sprintf("Health report sent (status: %s)", health["status"]))
	log.Info().Str("status", health["status"].(string)).Msg("cron: telegram_health done")
}

type healthData map[string]interface{}

func (s *Scheduler) gatherHealth() healthData {
	// DB size
	var dbSize float64
	s.db.Raw("SELECT ROUND(SUM(data_length + index_length) / 1024 / 1024, 2) FROM information_schema.tables WHERE table_schema = DATABASE()").Scan(&dbSize)

	var tableCount int64
	s.db.Raw("SELECT COUNT(*) FROM information_schema.tables WHERE table_schema = DATABASE()").Scan(&tableCount)

	// Active sessions
	var activeSessions int64
	s.db.Table("radacct").Where("acctstoptime IS NULL").Count(&activeSessions)

	// Pending overdue invoices
	var pendingInvoices int64
	s.db.Table("invoices").Where("status = ? AND dueDate < ?", "PENDING", time.Now()).Count(&pendingInvoices)

	// PPPoE users
	var totalUsers int64
	s.db.Model(&models.PppoeUser{}).Count(&totalUsers)
	var activeUsers int64
	s.db.Model(&models.PppoeUser{}).Where("status = ?", "active").Count(&activeUsers)

	// Determine status
	status := "healthy"
	var issues []string
	if dbSize > 5000 {
		status = "critical"
		issues = append(issues, "Database size > 5GB")
	} else if dbSize > 1000 {
		status = "warning"
		issues = append(issues, "Database size > 1GB")
	}
	if pendingInvoices > 50 {
		if status == "healthy" {
			status = "warning"
		}
		issues = append(issues, fmt.Sprintf("%d overdue invoices", pendingInvoices))
	}

	return healthData{
		"status":          status,
		"size":            fmt.Sprintf("%.2f MB", dbSize),
		"tables":          tableCount,
		"activeSessions":  activeSessions,
		"totalUsers":      totalUsers,
		"activeUsers":     activeUsers,
		"pendingInvoices": pendingInvoices,
		"issues":          strings.Join(issues, ", "),
	}
}

func formatHealthReport(h healthData) string {
	statusEmoji := "✅"
	switch h["status"] {
	case "warning":
		statusEmoji = "⚠️"
	case "critical":
		statusEmoji = "🔴"
	}

	report := fmt.Sprintf(`<b>%s System Health Report</b>
━━━━━━━━━━━━━━━
<b>Status:</b> %s %s
<b>Database Size:</b> %s
<b>Tables:</b> %d
<b>Active Sessions:</b> %d
<b>Total Users:</b> %d
<b>Active Users:</b> %d
<b>Pending Invoices:</b> %d
<b>Time:</b> %s`,
		statusEmoji,
		h["status"], statusEmoji,
		h["size"],
		h["tables"],
		h["activeSessions"],
		h["totalUsers"],
		h["activeUsers"],
		h["pendingInvoices"],
		time.Now().Format("2006-01-02 15:04:05 WIB"),
	)

	if issues, ok := h["issues"].(string); ok && issues != "" {
		report += fmt.Sprintf("\n<b>⚠️ Issues:</b> %s", issues)
	}

	return report
}

// registerTelegramBackupCron reads settings from DB and registers the backup cron.
// Called once at scheduler startup. If settings change, the server must be restarted
// or TriggerJob("telegram_backup") can be used for manual runs.
func (s *Scheduler) registerTelegramBackupCron() {
	var settings models.TelegramBackupSettings
	if err := s.db.Where("enabled = true").Order("createdAt desc").First(&settings).Error; err != nil {
		log.Info().Msg("cron: telegram_backup disabled (no enabled settings)")
		return
	}

	cronExpr := telegramScheduleToCron(settings.Schedule, settings.ScheduleTime)
	s.cron.AddFunc(cronExpr, s.jobTelegramBackup)
	log.Info().Str("schedule", settings.Schedule).Str("time", settings.ScheduleTime).Str("cron", cronExpr).
		Msg("cron: telegram_backup registered")
}

// ─── Helpers ──────────────────────────────────────────────────────────────────

func telegramMysqlDump(targetPath string) error {
	host, user, pass, dbname := "127.0.0.1", "root", "", "salfanet_radius"
	if databaseURL := os.Getenv("DATABASE_URL"); databaseURL != "" {
		// Parse DATABASE_URL
		if u, err := parseDatabaseURL(databaseURL); err == nil {
			host = u.host
			user = u.user
			pass = u.pass
			dbname = u.dbname
		}
	}
	shellCmd := fmt.Sprintf("mysqldump -h%s -u%s --no-tablespaces %s | gzip > %s", host, user, dbname, targetPath)
	cmd := exec.Command("sh", "-c", shellCmd)
	cmd.Env = append(os.Environ(), "MYSQL_PWD="+pass)
	out, err := cmd.CombinedOutput()
	if err != nil {
		return fmt.Errorf("%s — %w", strings.TrimSpace(string(out)), err)
	}
	return nil
}

type dbURL struct {
	host, user, pass, dbname string
}

func parseDatabaseURL(url string) (*dbURL, error) {
	// Simple parser for mysql://user:pass@host:port/dbname
	u := &dbURL{}
	if !strings.HasPrefix(url, "mysql://") && !strings.HasPrefix(url, "mysql+") {
		// Try generic URL parsing
	}
	// Remove scheme
	rest := url
	if idx := strings.Index(rest, "://"); idx >= 0 {
		rest = rest[idx+3:]
	}
	// Split user:pass@host:port/db
	atIdx := strings.LastIndex(rest, "@")
	if atIdx >= 0 {
		userPass := rest[:atIdx]
		rest = rest[atIdx+1:]
		colonIdx := strings.Index(userPass, ":")
		if colonIdx >= 0 {
			u.user = userPass[:colonIdx]
			u.pass = userPass[colonIdx+1:]
		} else {
			u.user = userPass
		}
	}
	// Split host:port/db
	slashIdx := strings.Index(rest, "/")
	if slashIdx >= 0 {
		u.host = rest[:slashIdx]
		u.dbname = rest[slashIdx+1:]
	} else {
		u.host = rest
	}
	// Remove query string from dbname
	if qIdx := strings.Index(u.dbname, "?"); qIdx >= 0 {
		u.dbname = u.dbname[:qIdx]
	}
	// Remove port from host
	if cIdx := strings.Index(u.host, ":"); cIdx >= 0 {
		u.host = u.host[:cIdx]
	}
	return u, nil
}

func telegramSendDocument(botToken, chatId, filePath, caption string) error {
	f, err := os.Open(filePath)
	if err != nil {
		return fmt.Errorf("open file: %w", err)
	}
	defer f.Close()

	var buf bytes.Buffer
	w := multipart.NewWriter(&buf)
	_ = w.WriteField("chat_id", chatId)
	if caption != "" {
		_ = w.WriteField("caption", caption)
	}
	fw, err := w.CreateFormFile("document", filepath.Base(filePath))
	if err != nil {
		return err
	}
	if _, err = io.Copy(fw, f); err != nil {
		return err
	}
	w.Close()

	apiURL := fmt.Sprintf("https://api.telegram.org/bot%s/sendDocument", botToken)
	resp, err := http.Post(apiURL, w.FormDataContentType(), &buf)
	if err != nil {
		return fmt.Errorf("sendDocument: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("sendDocument status %d: %s", resp.StatusCode, string(body))
	}
	return nil
}

func telegramSendMessage(botToken, chatId, text string) error {
	apiURL := fmt.Sprintf("https://api.telegram.org/bot%s/sendMessage", botToken)
	payload := fmt.Sprintf(`{"chat_id":%q,"text":%q,"parse_mode":"HTML"}`, chatId, text)
	resp, err := http.Post(apiURL, "application/json", strings.NewReader(payload))
	if err != nil {
		return fmt.Errorf("sendMessage: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("sendMessage status %d: %s", resp.StatusCode, string(body))
	}
	return nil
}

// telegramScheduleToCron converts TelegramBackupSettings schedule to cron expression.
func telegramScheduleToCron(schedule, scheduleTime string) string {
	parts := strings.Split(scheduleTime, ":")
	hour, minute := 2, 0
	if len(parts) == 2 {
		hour, _ = strconv.Atoi(parts[0])
		minute, _ = strconv.Atoi(parts[1])
	}

	switch schedule {
	case "daily":
		return fmt.Sprintf("0 %d %d * * *", minute, hour)
	case "12h":
		return fmt.Sprintf("0 %d %d,%d * * *", minute, hour, (hour+12)%24)
	case "6h":
		return fmt.Sprintf("0 %d %d,%d,%d,%d * * *", minute, hour, (hour+6)%24, (hour+12)%24, (hour+18)%24)
	case "weekly":
		return fmt.Sprintf("0 %d %d * * 0", minute, hour)
	default:
		return fmt.Sprintf("0 %d %d * * *", minute, hour)
	}
}

File: src/server/jobs/telegram-cron.ts
import 'server-only'
import cron, { ScheduledTask } from 'node-cron';
import { prisma } from '@/server/db/client';
import { nanoid } from 'nanoid';
import { sendBackupToTelegram, sendHealthReport } from '@/server/services/notifications/telegram.service';
import { createBackup } from '@/server/services/backup.service';
import * as fs from 'fs/promises';

let backupCronJob: ScheduledTask | null = null;
let healthCronJob: ScheduledTask | null = null;

// Mutex flags to prevent race-condition double-start
let backupCronStarting = false;
let healthCronStarting = false;

/**
 * Create and send database backup to Telegram
 */
export async function autoBackupToTelegram(): Promise<{ success: boolean; error?: string }> {
  const startedAt = new Date();

  // Deduplication guard: skip if another instance already started within the last 5 minutes.
  // This prevents double-send when multiple processes (e.g. runner.ts + cron-service) both
  // trigger backup at the same scheduled time.
  const fiveMinutesAgo = new Date(startedAt.getTime() - 5 * 60 * 1000);
  const recentRun = await prisma.cronHistory.findFirst({
    where: {
      jobType: 'telegram_backup',
      status: { in: ['running', 'success'] },
      startedAt: { gt: fiveMinutesAgo },
    },
    orderBy: { startedAt: 'desc' },
  });
  if (recentRun) {
    console.log(`[Telegram Backup] Skipping duplicate -- already ${recentRun.status} since ${recentRun.startedAt.toISOString()}`);
    return { success: true };
  }

  const history = await prisma.cronHistory.create({
    data: {
      id: nanoid(),
      jobType: 'telegram_backup',
      status: 'running',
      startedAt,
    },
  });

  try {
    // Get Telegram settings
    const settings = await prisma.telegramBackupSettings.findFirst({
      where: { enabled: true },
      orderBy: { createdAt: 'desc' },
    });

    if (!settings) {
      const completedAt = new Date();
      await prisma.cronHistory.update({
        where: { id: history.id },
        data: {
          status: 'success',
          completedAt,
          duration: completedAt.getTime() - startedAt.getTime(),
          result: 'Telegram backup disabled, skipped',
        },
      });
      return { success: true };
    }

    // Create backup
    console.log('[Telegram Backup] Creating database backup...');
    const backupResult = await createBackup('auto');
    
    if (!backupResult.success) {
      throw new Error('Backup creation failed');
    }

    const backup = backupResult.backup;

    // Check if file exists
    try {
      await fs.access(backupResult.filepath);
    } catch {
      throw new Error('Backup file not found on disk');
    }

    // Send to Telegram
    console.log('[Telegram Backup] Sending backup to Telegram...');
    const sendResult = await sendBackupToTelegram(
      {
        botToken: settings.botToken,
        chatId: settings.chatId,
        topicId: settings.backupTopicId || undefined,
      },
      backupResult.filepath,
      backup.filesize
    );

    if (!sendResult.success) {
      throw new Error(sendResult.error || 'Failed to send to Telegram');
    }

    // Clean up old backups if needed
    if (settings.keepLastN > 0) {
      const allBackups = await prisma.backupHistory.findMany({
        where: { type: 'auto' },
        orderBy: { createdAt: 'desc' },
      });

      const backupsToDelete = allBackups.slice(settings.keepLastN);
      
      for (const oldBackup of backupsToDelete) {
        try {
          if (oldBackup.filepath) {
            await fs.unlink(oldBackup.filepath);
          }
          await prisma.backupHistory.delete({ where: { id: oldBackup.id } });
          console.log(`[Telegram Backup] Deleted old backup: ${oldBackup.filename}`);
        } catch (error) {
          console.error(`[Telegram Backup] Failed to delete ${oldBackup.filename}:`, error);
        }
      }
    }

    const completedAt = new Date();
    await prisma.cronHistory.update({
      where: { id: history.id },
      data: {
        status: 'success',
        completedAt,
        duration: completedAt.getTime() - startedAt.getTime(),
        result: `Backup sent to Telegram: ${backup.filename}`,
      },
    });

    console.log('[Telegram Backup] Completed successfully');
    return { success: true };
  } catch (error: any) {
    console.error('[Telegram Backup] Error:', error);
    
    const completedAt = new Date();
    await prisma.cronHistory.update({
      where: { id: history.id },
      data: {
        status: 'error',
        completedAt,
        duration: completedAt.getTime() - startedAt.getTime(),
        error: error.message,
      },
    });

    return { success: false, error: error.message };
  }
}

/**
 * Send comprehensive health check to Telegram
 */
export async function sendHealthCheckToTelegram(): Promise<{ success: boolean; error?: string }> {
  const startedAt = new Date();
  
  const history = await prisma.cronHistory.create({
    data: {
      id: nanoid(),
      jobType: 'telegram_health',
      status: 'running',
      startedAt,
    },
  });

  try {
    // Get Telegram settings
    const settings = await prisma.telegramBackupSettings.findFirst({
      where: { enabled: true },
      orderBy: { createdAt: 'desc' },
    });

    if (!settings) {
      const completedAt = new Date();
      await prisma.cronHistory.update({
        where: { id: history.id },
        data: {
          status: 'success',
          completedAt,
          duration: completedAt.getTime() - startedAt.getTime(),
          result: 'Telegram health check disabled, skipped',
        },
      });
      return { success: true };
    }

    // Get comprehensive health data
    const health = await getComprehensiveHealth();

    // Send to Telegram
    console.log('[Telegram Health] Sending health report...');
    const sendResult = await sendHealthReport(
      {
        botToken: settings.botToken,
        chatId: settings.chatId,
        topicId: settings.healthTopicId || undefined,
      },
      health
    );

    if (!sendResult.success) {
      throw new Error(sendResult.error || 'Failed to send to Telegram');
    }

    const completedAt = new Date();
    await prisma.cronHistory.update({
      where: { id: history.id },
      data: {
        status: 'success',
        completedAt,
        duration: completedAt.getTime() - startedAt.getTime(),
        result: `Health report sent (status: ${health.status})`,
      },
    });

    console.log('[Telegram Health] Completed successfully');
    return { success: true };
  } catch (error: any) {
    console.error('[Telegram Health] Error:', error);
    
    const completedAt = new Date();
    await prisma.cronHistory.update({
      where: { id: history.id },
      data: {
        status: 'error',
        completedAt,
        duration: completedAt.getTime() - startedAt.getTime(),
        error: error.message,
      },
    });

    return { success: false, error: error.message };
  }
}

/**
 * Get comprehensive system health including DB, billing, and RADIUS
 */
async function getComprehensiveHealth() {
  try {
    // Database health
    const sizeResult: any = await prisma.$queryRawUnsafe(`
      SELECT 
        ROUND(SUM(data_length + index_length) / 1024 / 1024, 2) as size_mb
      FROM information_schema.TABLES 
      WHERE table_schema = DATABASE()
    `);

    const tableResult: any = await prisma.$queryRawUnsafe(`
      SELECT COUNT(*) as count 
      FROM information_schema.TABLES 
      WHERE table_schema = DATABASE()
    `);

    const connectionResult: any = await prisma.$queryRawUnsafe(`
      SHOW STATUS LIKE 'Threads_connected'
    `);

    const uptimeResult: any = await prisma.$queryRawUnsafe(`
      SHOW STATUS LIKE 'Uptime'
    `);

    // RADIUS health - check active sessions
    const activeSessions = await prisma.radacct.count({
      where: {
        acctstoptime: null,
      },
    });

    // Billing health - check pending invoices
    const pendingInvoices = await prisma.invoice.count({
      where: {
        status: 'PENDING',
        dueDate: { lt: new Date() },
      },
    });

    // PPPoE users count
    const totalUsers = await prisma.pppoeUser.count();
    const activeUsers = await prisma.pppoeUser.count({
      where: { status: 'ACTIVE' },
    });

    const sizeMB = sizeResult[0]?.size_mb || 0;
    const tableCount = Number(tableResult[0]?.count) || 0;
    const connections = connectionResult[0]?.Value || '0';
    const uptimeSeconds = Number(uptimeResult[0]?.Value) || 0;
    
    const days = Math.floor(uptimeSeconds / 86400);
    const hours = Math.floor((uptimeSeconds % 86400) / 3600);
    const uptime = `${days}d ${hours}h`;

    // Determine overall status
    let status = 'healthy';
    const issues = [];

    if (sizeMB > 5000) {
      status = 'critical';
      issues.push('Database size > 5GB');
    } else if (sizeMB > 1000) {
      status = 'warning';
      issues.push('Database size > 1GB');
    }

    if (Number(connections) > 100) {
      status = 'critical';
      issues.push('Too many DB connections');
    } else if (Number(connections) > 50) {
      if (status === 'healthy') status = 'warning';
      issues.push('High DB connections');
    }

    if (pendingInvoices > 50) {
      if (status === 'healthy') status = 'warning';
      issues.push(`${pendingInvoices} overdue invoices`);
    }

    return {
      status,
      size: `${sizeMB} MB`,
      tables: tableCount,
      connections: connections,
      uptime,
      activeSessions,
      totalUsers,
      activeUsers,
      pendingInvoices,
      issues: issues.length > 0 ? issues.join(', ') : undefined,
    };
  } catch (error) {
    console.error('[Health Check] Error:', error);
    return {
      status: 'unknown',
      size: 'N/A',
      tables: 0,
      connections: 'N/A',
      uptime: 'N/A',
      activeSessions: 0,
      totalUsers: 0,
      activeUsers: 0,
      pendingInvoices: 0,
    };
  }
}

/**
 * Start or restart backup cron job based on settings
 */
export async function startBackupCron() {
  try {
    // Prevent concurrent starts (race condition guard)
    if (backupCronStarting) {
      console.log('[Telegram Backup Cron] Already starting, skipping duplicate call');
      return;
    }
    backupCronStarting = true;

    // Stop existing job if running
    if (backupCronJob) {
      backupCronJob.stop();
      backupCronJob = null;
    }

    // Get settings
    const settings = await prisma.telegramBackupSettings.findFirst({
      where: { enabled: true },
      orderBy: { createdAt: 'desc' },
    });

    if (!settings) {
      console.log('[Telegram Backup Cron] Disabled or not configured, skipping');
      backupCronStarting = false;
      return;
    }

    console.log('[Telegram Backup Cron] Settings found:', {
      enabled: settings.enabled,
      schedule: settings.schedule,
      scheduleTime: settings.scheduleTime,
    });

    // Parse schedule time (HH:mm format)
    const [hour, minute] = settings.scheduleTime.split(':').map(Number);

  // Build cron expression based on schedule
  let cronExpression = '';
  
  switch (settings.schedule) {
    case 'daily':
      cronExpression = `${minute} ${hour} * * *`; // Daily at specified time
      break;
    case '12h':
      cronExpression = `${minute} ${hour},${(hour + 12) % 24} * * *`; // Every 12 hours
      break;
    case '6h':
      cronExpression = `${minute} ${hour},${(hour + 6) % 24},${(hour + 12) % 24},${(hour + 18) % 24} * * *`; // Every 6 hours
      break;
    case 'weekly':
      cronExpression = `${minute} ${hour} * * 0`; // Weekly on Sunday
      break;
    default:
      cronExpression = `${minute} ${hour} * * *`; // Default: daily
  }

  console.log(`[Telegram Backup Cron] Starting with schedule: ${settings.schedule} at ${settings.scheduleTime} WIB (cron: ${cronExpression})`);

  backupCronJob = cron.schedule(cronExpression, async () => {
    console.log('[Telegram Backup Cron] Running...');
    const result = await autoBackupToTelegram();
    console.log('[Telegram Backup Cron] Result:', result);
  }, {
    timezone: 'Asia/Jakarta' // WIB timezone
  });

  backupCronJob.start();
  console.log('[Telegram Backup Cron] Successfully started');
  } catch (error) {
    console.error('[Telegram Backup Cron] Failed to start:', error);
  } finally {
    backupCronStarting = false;
  }
}

/**
 * Start health check cron (runs every hour)
 */
export async function startHealthCron() {
  try {
    // Prevent concurrent starts (race condition guard)
    if (healthCronStarting) {
      console.log('[Telegram Health Cron] Already starting, skipping duplicate call');
      return;
    }
    healthCronStarting = true;

    // Stop existing job if running
    if (healthCronJob) {
      healthCronJob.stop();
      healthCronJob = null;
    }

    // Check if enabled
    const settings = await prisma.telegramBackupSettings.findFirst({
      where: { enabled: true },
      orderBy: { createdAt: 'desc' },
    });

    if (!settings) {
      console.log('[Telegram Health Cron] Disabled or not configured, skipping');
      healthCronStarting = false;
      return;
    }

    console.log('[Telegram Health Cron] Starting (every hour at :00)');

    healthCronJob = cron.schedule('0 * * * *', async () => {
      console.log('[Telegram Health Cron] Running...');
      const result = await sendHealthCheckToTelegram();
      console.log('[Telegram Health Cron] Result:', result);
    }, {
      timezone: 'Asia/Jakarta' // WIB timezone
    });

    healthCronJob.start();
    console.log('[Telegram Health Cron] Successfully started');
  } catch (error) {
    console.error('[Telegram Health Cron] Failed to start:', error);
  } finally {
    healthCronStarting = false;
  }
}

/**
 * Stop backup cron
 */
export function stopBackupCron() {
  if (backupCronJob) {
    backupCronJob.stop();
    backupCronJob = null;
    console.log('[Telegram Backup Cron] Stopped');
  }
}

/**
 * Stop health cron
 */
export function stopHealthCron() {
  if (healthCronJob) {
    healthCronJob.stop();
    healthCronJob = null;
    console.log('[Telegram Health Cron] Stopped');
  }
}

/**
 * Get cron status
 */
export function getTelegramCronStatus() {
  return {
    backup: {
      running: backupCronJob !== null,
    },
    health: {
      running: healthCronJob !== null,
    },
  };
}
